File: bindings/react-native/src/facets/telemetry.ts
import UldrenLoom from '../NativeUldrenLoom';
import { authArgs, keyArgs } from '../internal';
import type { LoomAuth, LoomKey } from '../internal';

export function metricsPutDescriptor(
  loomPath: string,
  workspace: string,
  descriptor: Uint8Array | number[],
  loomKey?: LoomKey,
  auth?: LoomAuth
): Promise<void> {
  const [passphrase, kek] = keyArgs(loomKey);
  const [authPrincipal, authPassphrase] = authArgs(auth);
  return UldrenLoom.metricsPutDescriptor(
    loomPath,
    workspace,
    Array.from(descriptor),
    passphrase,
    kek,
    authPrincipal,
    authPassphrase
  );
}

export async function metricsGetDescriptor(
  loomPath: string,
  workspace: string,
  name: string,
  loomKey?: LoomKey,
  auth?: LoomAuth
): Promise<Uint8Array | null> {
  const [passphrase, kek] = keyArgs(loomKey);
  const [authPrincipal, authPassphrase] = authArgs(auth);
  const bytes = await UldrenLoom.metricsGetDescriptor(
    loomPath,
    workspace,
    name,
    passphrase,
    kek,
    authPrincipal,
    authPassphrase
  );
  return bytes == null ? null : Uint8Array.from(bytes);
}

export function metricsPutObservation(
  loomPath: string,
  workspace: string,
  descriptorName: string,
  observation: Uint8Array | number[],
  loomKey?: LoomKey,
  auth?: LoomAuth
): Promise<void> {
  const [passphrase, kek] = keyArgs(loomKey);
  const [authPrincipal, authPassphrase] = authArgs(auth);
  return UldrenLoom.metricsPutObservation(
    loomPath,
    workspace,
    descriptorName,
    Array.from(observation),
    passphrase,
    kek,
    authPrincipal,
    authPassphrase
  );
}

export async function metricsQuery(
  loomPath: string,
  workspace: string,
  descriptorName: string,
  fromTimestampMs: number | bigint | string,
  toTimestampMs: number | bigint | string,
  maxSeries: number,
  maxGroups: number,
  maxSamples: number,
  maxOutputBytes: number | bigint | string,
  nowTimestampMs: number | bigint | string,
  loomKey?: LoomKey,
  auth?: LoomAuth
): Promise<Uint8Array> {
  const [passphrase, kek] = keyArgs(loomKey);
  const [authPrincipal, authPassphrase] = authArgs(auth);
  const bytes = await UldrenLoom.metricsQuery(
    loomPath,
    workspace,
    descriptorName,
    String(fromTimestampMs),
    String(toTimestampMs),
    maxSeries,
    maxGroups,
    maxSamples,
    String(maxOutputBytes),
    String(nowTimestampMs),
    passphrase,
    kek,
    authPrincipal,
    authPassphrase
  );
  return Uint8Array.from(bytes);
}

export function logsPutRecord(
  loomPath: string,
  workspace: string,
  record: Uint8Array | number[],
  loomKey?: LoomKey,
  auth?: LoomAuth
): Promise<string> {
  const [passphrase, kek] = keyArgs(loomKey);
  const [authPrincipal, authPassphrase] = authArgs(auth);
  return UldrenLoom.logsPutRecord(
    loomPath,
    workspace,
    Array.from(record),
    passphrase,
    kek,
    authPrincipal,
    authPassphrase
  );
}

export async function logsGetRecord(
  loomPath: string,
  workspace: string,
  recordId: string,
  loomKey?: LoomKey,
  auth?: LoomAuth
): Promise<Uint8Array | null> {
  const [passphrase, kek] = keyArgs(loomKey);
  const [authPrincipal, authPassphrase] = authArgs(auth);
  const bytes = await UldrenLoom.logsGetRecord(
    loomPath,
    workspace,
    recordId,
    passphrase,
    kek,
    authPrincipal,
    authPassphrase
  );
  return bytes == null ? null : Uint8Array.from(bytes);
}

export async function logsQuery(
  loomPath: string,
  workspace: string,
  fromTimeUnixNano: number | bigint | string,
  toTimeUnixNano: number | bigint | string,
  maxRecords: number,
  maxOutputBytes: number | bigint | string,
  loomKey?: LoomKey,
  auth?: LoomAuth
): Promise<Uint8Array> {
  const [passphrase, kek] = keyArgs(loomKey);
  const [authPrincipal, authPassphrase] = authArgs(auth);
  const bytes = await UldrenLoom.logsQuery(
    loomPath,
    workspace,
    String(fromTimeUnixNano),
    String(toTimeUnixNano),
    maxRecords,
    String(maxOutputBytes),
    passphrase,
    kek,
    authPrincipal,
    authPassphrase
  );
  return Uint8Array.from(bytes);
}

export function tracesPutSpan(
  loomPath: string,
  workspace: string,
  span: Uint8Array | number[],
  loomKey?: LoomKey,
  auth?: LoomAuth
): Promise<void> {
  const [passphrase, kek] = keyArgs(loomKey);
  const [authPrincipal, authPassphrase] = authArgs(auth);
  return UldrenLoom.tracesPutSpan(
    loomPath,
    workspace,
    Array.from(span),
    passphrase,
    kek,
    authPrincipal,
    authPassphrase
  );
}

export async function tracesGetSpan(
  loomPath: string,
  workspace: string,
  traceId: string,
  spanId: string,
  loomKey?: LoomKey,
  auth?: LoomAuth
): Promise<Uint8Array | null> {
  const [passphrase, kek] = keyArgs(loomKey);
  const [authPrincipal, authPassphrase] = authArgs(auth);
  const bytes = await UldrenLoom.tracesGetSpan(
    loomPath,
    workspace,
    traceId,
    spanId,
    passphrase,
    kek,
    authPrincipal,
    authPassphrase
  );
  return bytes == null ? null : Uint8Array.from(bytes);
}

export async function tracesTraceSpans(
  loomPath: string,
  workspace: string,
  traceId: string,
  maxSpans: number,
  maxOutputBytes: number | bigint | string,
  loomKey?: LoomKey,
  auth?: LoomAuth
): Promise<Uint8Array> {
  const [passphrase, kek] = keyArgs(loomKey);
  const [authPrincipal, authPassphrase] = authArgs(auth);
  const bytes = await UldrenLoom.tracesTraceSpans(
    loomPath,
    workspace,
    traceId,
    maxSpans,
    String(maxOutputBytes),
    passphrase,
    kek,
    authPrincipal,
    authPassphrase
  );
  return Uint8Array.from(bytes);
}

export async function tracesQuery(
  loomPath: string,
  workspace: string,
  fromStartTimeNs: number | bigint | string,
  toStartTimeNs: number | bigint | string,
  maxSpans: number,
  maxOutputBytes: number | bigint | string,
  loomKey?: LoomKey,
  auth?: LoomAuth
): Promise<Uint8Array> {
  const [passphrase, kek] = keyArgs(loomKey);
  const [authPrincipal, authPassphrase] = authArgs(auth);
  const bytes = await UldrenLoom.tracesQuery(
    loomPath,
    workspace,
    String(fromStartTimeNs),
    String(toStartTimeNs),
    maxSpans,
    String(maxOutputBytes),
    passphrase,
    kek,
    authPrincipal,
    authPassphrase
  );
  return Uint8Array.from(bytes);
}

File: bindings/android/src/commonMain/kotlin/ai/uldren/loom/facets/telemetry.kt
package ai.uldren.loom

expect fun Loom.metricsPutDescriptor(
    path: String,
    workspace: String,
    descriptor: ByteArray,
    passphrase: String? = null,
    kek: ByteArray? = null,
    authPrincipal: String? = null,
    authPassphrase: String? = null,
)

expect fun Loom.metricsGetDescriptor(
    path: String,
    workspace: String,
    name: String,
    passphrase: String? = null,
    kek: ByteArray? = null,
    authPrincipal: String? = null,
    authPassphrase: String? = null,
): ByteArray?

expect fun Loom.metricsPutObservation(
    path: String,
    workspace: String,
    descriptorName: String,
    observation: ByteArray,
    passphrase: String? = null,
    kek: ByteArray? = null,
    authPrincipal: String? = null,
    authPassphrase: String? = null,
)

expect fun Loom.metricsQuery(
    path: String,
    workspace: String,
    descriptorName: String,
    fromTimestampMs: Long,
    toTimestampMs: Long,
    maxSeries: Int,
    maxGroups: Int,
    maxSamples: Int,
    maxOutputBytes: Long,
    nowTimestampMs: Long,
    passphrase: String? = null,
    kek: ByteArray? = null,
    authPrincipal: String? = null,
    authPassphrase: String? = null,
): ByteArray

expect fun Loom.logsPutRecord(
    path: String,
    workspace: String,
    record: ByteArray,
    passphrase: String? = null,
    kek: ByteArray? = null,
    authPrincipal: String? = null,
    authPassphrase: String? = null,
): String

expect fun Loom.logsGetRecord(
    path: String,
    workspace: String,
    recordId: String,
    passphrase: String? = null,
    kek: ByteArray? = null,
    authPrincipal: String? = null,
    authPassphrase: String? = null,
): ByteArray?

expect fun Loom.logsQuery(
    path: String,
    workspace: String,
    fromTimeUnixNano: Long,
    toTimeUnixNano: Long,
    maxRecords: Int,
    maxOutputBytes: Long,
    passphrase: String? = null,
    kek: ByteArray? = null,
    authPrincipal: String? = null,
    authPassphrase: String? = null,
): ByteArray

expect fun Loom.tracesPutSpan(
    path: String,
    workspace: String,
    span: ByteArray,
    passphrase: String? = null,
    kek: ByteArray? = null,
    authPrincipal: String? = null,
    authPassphrase: String? = null,
)

expect fun Loom.tracesGetSpan(
    path: String,
    workspace: String,
    traceId: String,
    spanId: String,
    passphrase: String? = null,
    kek: ByteArray? = null,
    authPrincipal: String? = null,
    authPassphrase: String? = null,
): ByteArray?

expect fun Loom.tracesTraceSpans(
    path: String,
    workspace: String,
    traceId: String,
    maxSpans: Int,
    maxOutputBytes: Long,
    passphrase: String? = null,
    kek: ByteArray? = null,
    authPrincipal: String? = null,
    authPassphrase: String? = null,
): ByteArray

expect fun Loom.tracesQuery(
    path: String,
    workspace: String,
    fromStartTimeNs: Long,
    toStartTimeNs: Long,
    maxSpans: Int,
    maxOutputBytes: Long,
    passphrase: String? = null,
    kek: ByteArray? = null,
    authPrincipal: String? = null,
    authPassphrase: String? = null,
): ByteArray

class TelemetryOps(private val s: LoomSession) {
    fun metricsPutDescriptor(workspace: String, descriptor: ByteArray) =
        Loom.metricsPutDescriptor(s.path, workspace, descriptor, s.passphrase, s.kek, s.authPrincipal, s.authPassphrase)

    fun metricsGetDescriptor(workspace: String, name: String): ByteArray? =
        Loom.metricsGetDescriptor(s.path, workspace, name, s.passphrase, s.kek, s.authPrincipal, s.authPassphrase)

    fun metricsPutObservation(workspace: String, descriptorName: String, observation: ByteArray) =
        Loom.metricsPutObservation(
            s.path,
            workspace,
            descriptorName,
            observation,
            s.passphrase,
            s.kek,
            s.authPrincipal,
            s.authPassphrase,
        )

    fun metricsQuery(
        workspace: String,
        descriptorName: String,
        fromTimestampMs: Long,
        toTimestampMs: Long,
        maxSeries: Int,
        maxGroups: Int,
        maxSamples: Int,
        maxOutputBytes: Long,
        nowTimestampMs: Long,
    ): ByteArray = Loom.metricsQuery(
        s.path,
        workspace,
        descriptorName,
        fromTimestampMs,
        toTimestampMs,
        maxSeries,
        maxGroups,
        maxSamples,
        maxOutputBytes,
        nowTimestampMs,
        s.passphrase,
        s.kek,
        s.authPrincipal,
        s.authPassphrase,
    )

    fun logsPutRecord(workspace: String, record: ByteArray): String =
        Loom.logsPutRecord(s.path, workspace, record, s.passphrase, s.kek, s.authPrincipal, s.authPassphrase)

    fun logsGetRecord(workspace: String, recordId: String): ByteArray? =
        Loom.logsGetRecord(s.path, workspace, recordId, s.passphrase, s.kek, s.authPrincipal, s.authPassphrase)

    fun logsQuery(
        workspace: String,
        fromTimeUnixNano: Long,
        toTimeUnixNano: Long,
        maxRecords: Int,
        maxOutputBytes: Long,
    ): ByteArray = Loom.logsQuery(
        s.path,
        workspace,
        fromTimeUnixNano,
        toTimeUnixNano,
        maxRecords,
        maxOutputBytes,
        s.passphrase,
        s.kek,
        s.authPrincipal,
        s.authPassphrase,
    )

    fun tracesPutSpan(workspace: String, span: ByteArray) =
        Loom.tracesPutSpan(s.path, workspace, span, s.passphrase, s.kek, s.authPrincipal, s.authPassphrase)

    fun tracesGetSpan(workspace: String, traceId: String, spanId: String): ByteArray? =
        Loom.tracesGetSpan(s.path, workspace, traceId, spanId, s.passphrase, s.kek, s.authPrincipal, s.authPassphrase)

    fun tracesTraceSpans(workspace: String, traceId: String, maxSpans: Int, maxOutputBytes: Long): ByteArray =
        Loom.tracesTraceSpans(
            s.path,
            workspace,
            traceId,
            maxSpans,
            maxOutputBytes,
            s.passphrase,
            s.kek,
            s.authPrincipal,
            s.authPassphrase,
        )

    fun tracesQuery(
        workspace: String,
        fromStartTimeNs: Long,
        toStartTimeNs: Long,
        maxSpans: Int,
        maxOutputBytes: Long,
    ): ByteArray = Loom.tracesQuery(
        s.path,
        workspace,
        fromStartTimeNs,
        toStartTimeNs,
        maxSpans,
        maxOutputBytes,
        s.passphrase,
        s.kek,
        s.authPrincipal,
        s.authPassphrase,
    )
}
